Refactored Plate Replication functionality added to Metapool

charles-cowart opened the pull request:
Initial migrating of Rodolfo's new notebook functions to metapool. Minimal changes made to support the new notebook in development.

antgonza commented:
Can you add tests before I review?

charles-cowart commented:
@antgonza Ready for review. The code that supports each test is a little repetitive, and the metadata files could be named better.

antgonza commented on metapool/plate.py:
```diff
@@ -412,3 +413,287 @@ def _autopool_norm(pool_failures, total_nmol, min_conc, sample_concs,
+        # since all values in self.d are generated in the same order by
+        # the same function and saved as OrderedDicts, the 384-well locations
+        # from any src quadrant will map directly to those of the dst quad.
+        src = list(self.map_to_384[str(src_quad)].values())
+        dst = list(self.map_to_384[str(dst_quad)].values())
```
I'm not sure the list conversion is necessary but it might be for map, do you know?

antgonza commented on metapool/plate.py:
```diff
@@ -412,3 +413,287 @@ def _autopool_norm(pool_failures, total_nmol, min_conc, sample_concs,
+        '''
+        Check if one or more 384-well plate IDs are valid.
+        :param locations: A list of one or more locations
+        :return: A list of valid locations or None if none were valid.
```
This goes back to my previous comment, is it really safe to return None vs. an empty array?

charles-cowart replied:
I can go either way. I included the functionality for the wetlab, but I'm not sure if they'll still need to use it since PlateReplicant handles all of the checking for them. I agree it's safer for them to get a list returned.

antgonza commented on metapool/tests/test_plate.py:
```diff
@@ -563,6 +564,140 @@ def test_sample_plate_has_no_spaces(self):
+
+        pd.set_option('display.max_columns', None)
+
+        obs = obs.set_index('Sample').to_dict(orient='index')
+        exp = exp.set_index('Sample').to_dict(orient='index')
+
+        sample_names = set(obs.keys()).union(set(exp.keys()))
+
+        for sample_name in sample_names:
+            self.assertDictEqual(obs[sample_name], exp[sample_name])
+
+        # replicate a valid source to empty sources 2 and 4 plus overwriting
+        # source 3 with overwrites not allowed. Should return an Error.
+
+        with self.assertRaisesRegex(ValueError, 'Quadrant 3 is a source '
+                                                'quadrant'):
+            pr.make_replicates(input_df, {1: [2, 3, 4]}, overwrite=False)
```
Is there a reason to not use here pandas.testing.assert_series_equal ?

antgonza replied:
This also applies to all the tests in this file.

charles-cowart replied:
Ty for the tip! I ended up using assert_frame_equal().

antgonza reviewed (changes requested):
Looking good but a few questions.

charles-cowart commented on metapool/plate.py:
```diff
@@ -412,3 +413,287 @@ def _autopool_norm(pool_failures, total_nmol, min_conc, sample_concs,
+    STATUS_SOURCE = 'source'
+    STATUS_DESTINATION = 'destination'
+
+    row_letters = [chr(ord('A') + x) for x in range(0, 16)]
```
Nice! I didn't know about this, thanks!

charles-cowart commented on metapool/plate.py:
```diff
@@ -412,3 +413,287 @@ def _autopool_norm(pool_failures, total_nmol, min_conc, sample_concs,
+        # this allows us to store output dataframes for each quadrant,
+        # overwrite them on demand, and save final concatenation for when
+        # we're done.
+        self.data = {'1': None, '2': None, '3': None, '4': None}
```
Since there's three valid states a quad can be in (empty, destination, source) we keep track of it separately in self.state. State is checked before we read from/write to self.data so for the most part it's okay. The only time we iterate through self.data keys is when we concatenate the four quadrants at the end. I'll revise this to go by self.state instead, thanks for catching it! Based on discussions with Rodolfo and the earlier code I assumed all four quadrants would be represented.

(I'm not sure I prefer the separation, but the only alternative I can see would be to make self.data values be another dict that stores the state under one key and the data frame under another.)

charles-cowart commented on metapool/plate.py:
```diff
@@ -412,3 +413,287 @@ def _autopool_norm(pool_failures, total_nmol, min_conc, sample_concs,
+    row_letters = [chr(ord('A') + x) for x in range(0, 16)]
+
+    # aka ['blue', 'green', 'red', 'yellow']
+    quadrants = ['1', '2', '3', '4']
```
I prefer the colors as well, but Rodolfo told me the lab standarderized on this so I made the switch.

antgonza commented on metapool/tests/test_plate.py:
```diff
@@ -563,6 +565,110 @@ def test_sample_plate_has_no_spaces(self):
+        self.input_df = pd.read_csv('metapool/tests/data/input_plate.tsv',
+                                    sep='\t', dtype=str)
+
+    def _frames_equal(self, obs, exp):
```
Well, I do _not_ think this is necessary as if the frames are equal, they are simply equal and there is nothing to test against None but if they are not the error is raised. Another way to say this, nowhere (that I can tell) the code below actually testes for `self.assertFalse(_frames_equal....`. Can you simply use `assert_frame_equal`?

charles-cowart replied:
I suppose you're right. I confirmed that AssertionError raised by assert_frame_equal is a standard Error and that the standard self.assert*() methods also return the same type.

antgonza reviewed (commented):
Much cleaner, thank you. One big question and out of curiosity, is the notebook actually tested? Does it need any changes after the refactor?

antgonza commented on metapool/tests/test_plate.py:
```diff
@@ -563,6 +565,109 @@ def test_sample_plate_has_no_spaces(self):
+        exp = pd.read_csv('metapool/tests/data/file1.tsv',
+                          sep='\t', dtype=str)
+
+        obs = obs.set_index('Sample').sort_index()
+        exp = exp.set_index('Sample').sort_index()
+
```
Do we need to set_index/sort even if we are using `assert_frame_equal`? I would think no but not sure. This also applies for the rest of the tests.

charles-cowart replied:
Confirmed we do need, unless we use the check_like flag. I replaced the sort()s with the flag on all of the tests, ty!

charles-cowart commented:
@antgonza Eh, I cleaned up the notebook some - I removed the functionality that was no longer needed and replaced it with a call to the new PlateReplication() code. It's very complete and very well commented but references a specific project a lot. I didn't include the needed data files as one of them was very large. I think we could do with some smaller sample-files but I'll need to work with @RodolfoSalido or @mmbryant23 on that.

antgonza reviewed (approved):
After offline discussion we agreed on this version.

RodolfoSalido commented:
Just tested this and looks good. However, we didn't add the functionality of adding the 'contains_replicates' flag to the sample_sheet for the SPP to parse it. We had agreed on this when we were discussing the scope of this effort. 
![Screen Shot 2023-06-23 at 12 43 18 PM](https://github.com/biocore/metagenomics_pooling_notebook/assets/13193891/ea994ffe-09b4-489c-8eeb-b72673ec8a5f)
The functions that would need some changes are likely make_sample_sheet() and validate_and_scrub_sample_sheet. I have added another key to the Bioinformatics dictionary: 'contains_replicates' that gets a boolean True/False. 
![Screen Shot 2023-06-23 at 12 44 39 PM](https://github.com/biocore/metagenomics_pooling_notebook/assets/13193891/dea37910-44b7-475d-bd81-13a6b482b83b)

Should I raise this as a new issue ?

antgonza commented:
Yes, please open a new issue. Thank you for confirming that it _almost_ works as expected.
